Fix wrapper subclass serialization with custom sizes / strides

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137126
* #137125
* #137124
* #137099
* #137031
* __->__ #137030

Fixes #130154

This PR takes the strategy outlined in the above issue and clears out any cached sizes / strides PyCapsules before serialization. This affects the default subclass serialization logic.

The PyCapsule issue also affects `deepcopy`, so that's fixed here as well.

Note: I originally tried utilizing a context manager to remove / restore cached PyCapsules after serialization, but in practice the state returned from `_reduce_ex_internal()` references the actual `tensor.__dict__()`, so the problem persists once the cached values are restored. Instead, we have to be careful to remove the cached values in the right place so they're not re-cached when pulling out size / stride information for serialization.

jbschlosser committed:
Fix wrapper subclass serialization with custom sizes / strides

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix wrapper subclass serialization with custom sizes / strides"


Fixes #130154

This PR takes the strategy outlined in the above issue and clears out any cached sizes / strides PyCapsules before serialization. This affects the default subclass serialization logic.

The PyCapsule issue also affects `deepcopy`, so that's fixed here as well.

Note: I originally tried utilizing a context manager to remove / restore cached PyCapsules after serialization, but in practice the state returned from `_reduce_ex_internal()` references the actual `tensor.__dict__()`, so the problem persists once the cached values are restored. Instead, we have to be careful to remove the cached values in the right place so they're not re-cached when pulling out size / stride information for serialization.

[ghstack-poisoned]

albanD commented on torch/_tensor.py:
```diff
@@ -227,6 +241,9 @@ def __reduce_ex__(self, proto):
         func, args = self._reduce_ex_internal(proto)
+        # sizes / strides cache needs to be cleared here because it'll just be recached
+        # if cleared earlier. Note that state references the -actual- tensor dict.
+        _clear_cached_sizes_strides(self)
```
Should we make this a method on the class that user can also override? Or we expect them to override `__reduce_ex__` to do cleanup before they call `super().__reduce_ex__(` ?

jbschlosser replied:
Hm I like the idea of making this an overrideable class method. Is the name good or should it be generalized some to cover the general concept of clearing non-serializable cached stuff, wdyt?

jbschlosser commented on torch/testing/_internal/common_subclass.py:
```diff
@@ -196,6 +254,16 @@ def __init__(self, name, create_fn, closed_under_ops=True):
+def _create_and_access_shape(cls, shape):
+    sub = cls(torch.randn(shape))
+    # NB: Wrapper subclasses with custom dispatched sizes / strides cache this info
+    # on the first call via non-serializable PyCapsules.
```
will do, it's also for deepcopy tests :)

jbschlosser committed:
Update on "Fix wrapper subclass serialization with custom sizes / strides"


Fixes #130154

This PR takes the strategy outlined in the above issue and clears out any cached sizes / strides PyCapsules before serialization. This affects the default subclass serialization logic.

The PyCapsule issue also affects `deepcopy`, so that's fixed here as well.

Note: I originally tried utilizing a context manager to remove / restore cached PyCapsules after serialization, but in practice the state returned from `_reduce_ex_internal()` references the actual `tensor.__dict__()`, so the problem persists once the cached values are restored. Instead, we have to be careful to remove the cached values in the right place so they're not re-cached when pulling out size / stride information for serialization.

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix wrapper subclass serialization with custom sizes / strides"


Fixes #130154

This PR takes the strategy outlined in the above issue and clears out any cached sizes / strides PyCapsules before serialization. This affects the default subclass serialization logic.

The PyCapsule issue also affects `deepcopy`, so that's fixed here as well.

Note: I originally tried utilizing a context manager to remove / restore cached PyCapsules after serialization, but in practice the state returned from `_reduce_ex_internal()` references the actual `tensor.__dict__()`, so the problem persists once the cached values are restored. Instead, we have to be careful to remove the cached values in the right place so they're not re-cached when pulling out size / stride information for serialization.

[ghstack-poisoned]

albanD commented on torch/_tensor.py:
```diff
@@ -78,6 +78,22 @@ def _rebuild_from_type_v2(func, new_type, args, state):
 # torch/_C/__init__.pyi.in to add a type annotation for your method;
 # otherwise, it will not show up in autocomplete.
 class Tensor(torch._C.TensorBase):
+    def _clear_non_serializable_data(self):
```
Please add a docstring.
In particular mentioning that a subclass that overrides this  MUST call the super() variant

albanD replied:
Given that this is for content that is valid to clear, I think `cached` should be in the name

jbschlosser committed:
Update on "Fix wrapper subclass serialization with custom sizes / strides"


Fixes #130154

This PR takes the strategy outlined in the above issue and clears out any cached sizes / strides PyCapsules before serialization. This affects the default subclass serialization logic.

The PyCapsule issue also affects `deepcopy`, so that's fixed here as well.

Note: I originally tried utilizing a context manager to remove / restore cached PyCapsules after serialization, but in practice the state returned from `_reduce_ex_internal()` references the actual `tensor.__dict__()`, so the problem persists once the cached values are restored. Instead, we have to be careful to remove the cached values in the right place so they're not re-cached when pulling out size / stride information for serialization.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM !

jbschlosser commented:
@pytorchbot merge
